### PR TITLE
Core/Movement: Implement move time skipped handler

### PR DESCRIPTION
**Pull request by chaodhib**
Implement CMSG_MOVE_TIME_SKIPPED handler and move it to MovementHandler.cpp

**Changes proposed:**

The way the MOVE_TIME_SKIPPED packet works is as follow:
- The client sends it when the player [1] is moving but the client froze for a certain period of time.
- The packet only contains two information: the GUID of the moving unit. And the time elasped since the last movement packet has been sent.
- The meaning of the packet from the point of the view of the client that sends it is basically: I didn't run the physical simulation of the unit that I control for a certain period of time. Please, just play at a later time the last movement that was sent (new timestamp = previous timestamp + "time skipped").
- the server upon receiving that packet should 
	* retransmit the packet to the other clients that have a unit nearby.
	* update the movement timestamp of that unit (new timestamp = previous timestamp + "time skipped").
- the clients upon receiving also update the movement timestamp of that unit (new timestamp = previous timestamp + "time skipped") so that the previous move sent is "played" at a later time.

[1] or more generaly the unit currently moved by that client: it can be another player (in case of mind control for example), a vehicle, ...

More information on the packet exchange can be found in [this sniff extract](https://gist.github.com/chaodhib/e66010a8ecb77b406ff86f6bbe93a5e0#file-move-time-skipped-sniff). In it, we can see 3 packets sent by the client. The unit is moving between the first and third packet (and also before the first and after the third but that's not important here). The distance traveled between the first and third packet is only 500ms * run speed. Yet, the time difference between the 1st & 3rd packet is 500ms + 86ms. 86ms is the "time skipped". This confirms that the client didn't run the physical simulation during that duration.

**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [ ] master

**Issues addressed:** This is retail like. The implementation of that packet handler was missing. The movement should feel more smooth. Particularly when the [anti jitter buffer](https://trinitycore.atlassian.net/wiki/spaces/tc/pages/721256449/Movement) in the client will be utilized by TC, which is not the case at the moment.

**Tests performed:** (Does it build, tested in-game, etc.)
Tested in game.

**Known issues and TODO list:** 
The line `ASSERT(mover != nullptr);` could cause problems if mover control change (client A looses control of player B. Or client A gains control of Vehicle C, ...) is badly implemented. I didn't test these edge cases enough.

### CODE REVIEW

File: src/server/game/Handlers/MovementHandler.cpp [hunk 1]
```diff
@@ -608,3 +608,27 @@ void WorldSession::HandleSummonResponseOpcode(WorldPacket& recvData)
...
+{
+    TC_LOG_DEBUG("network", "WORLD: Received CMSG_MOVE_TIME_SKIPPED");
+
+    ObjectGuid guid;
+    uint32 timeSkipped;
+    recvData >> guid.ReadAsPacked();
+    recvData >> timeSkipped;
+
+    Unit* mover = GetPlayer()->m_unitMovedByMe;
+    ASSERT(mover != nullptr);                      // there must always be a mover
```
Comment (jackpoz) [inline src/server/game/Handlers/MovementHandler.cpp]: can't this just be a null check ? why does it have to be an assert ?

Comment (Shauren) [reply src/server/game/Handlers/MovementHandler.cpp]: You must NEVER assert in client packet handlers

Comment (chaodhib) [reply src/server/game/Handlers/MovementHandler.cpp]: Indeed. Good point. In that case though, existing code will need to be modified as well since there already are a few asserts in MovementHandler.

Comment (jackpoz) [reply src/server/game/Handlers/MovementHandler.cpp]: half of "crashes" we get seems to be triggered asserts :P

Comment (kelno) [reply src/server/game/Handlers/MovementHandler.cpp]: Well that's a good thing, they do outline abnormal behavior instead of failing silently

Comment (chaodhib) [reply src/server/game/Handlers/MovementHandler.cpp]: @kelno well yes and no. the issue is that an attacker could, by sending custom packets, intentionally crash a server running on TC. Example here: If a client sends a CMSG_MOVE_TIME_SKIPPED packet while his character is feared, the condition `mover != nullptr` won't be met and the server will crash.

Comment (kelno) [reply src/server/game/Handlers/MovementHandler.cpp]: Yes in this case it's totally bad I agree, I was making a more general comment in response for the last comment!

**Comment by chaodhib** [comment]
Done. Anything else before merging?

**Comment by jackpoz** [comment]
Looks good to me

**Comment by jackpoz** [comment]
Thanks for the PR :)
